src/gui/window.h

```diff
@@ -0,0 +1,102 @@
+/*
+Minetest
+Copyright (C) 2023 v-rob, Vincent Robinson <robinsonvincent89@gmail.com>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#pragma once
+
+#include "irrlichttypes_extrabloated.h"
+#include "gui/elem.h"
+#include "util/basic_macros.h"
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace ui
+{
+	// Serialized enum; do not change order of entries.
+	enum class WindowType
+	{
+		BG,
+		MASK,
+		HUD,
+		MESSAGE,
+		GUI,
+		FG,
```

Comment:
Idea: message -> chat to make the name less general and more obvious what it means

Reply:
Another note: I am thinking about making it possible for users to interact with `message` using a touchscreen, but not `hud`, so you can e.g. hit an X button to close a message.

I'm still deciding whether it's a good idea to add another window type for custom touchscreen controls, however, one that can't be hidden like the chat and only shows up on devices with touchscreens.

Reply:
> I'm still deciding whether it's a good idea to add another window type for custom touchscreen controls, however, one that can't be hidden like the chat and only shows up on devices with touchscreens.

I think that's something for later since it'll need to interface with the existing touch controls to be useful.

Reply:
It definitely is for later.  Plus, if we have custom controls, we probably also want custom keybindings for feature parity on desktop and mobile.